### ISSUE TITLE
[문제] GLM-4-9B LoRA 학습 테스트-1113

### ISSUE THREAD

**Issue by ryangkyung-thaki** [ThakiCloud/LLaMA-Factory#9]
```
### LoRA 설정
lora_rank: 8  ← 병목 가능성 높음
lora_target: all

### 학습 설정
per_device_train_batch_size: 32  ← 생각보다 큼!
gradient_accumulation_steps: 1
learning_rate: 1.0e-5  ← 가장 마지막 실험값

### 스케줄러
lr_scheduler_type: cosine
warmup_ratio: 0.1
```

🎯 개선 실험 제안
앞서 분석한 내용을 바탕으로, 우선순위대로 실험해볼 설정들입니다:
실험 1: LoRA Rank 증가 (최우선 ⭐⭐⭐)
methodlora_rank: 16 ( 8 → 16, 2배)
lora_alpha: 32  (rank의 2배)
learning_rate: 1.0e-4
이유:
Rank 8은 복잡한 reasoning에 부족할 수 있음
1.0e-4와 1.0e-5에서 동일한 결과 → 표현력 부족 가능성

실험 2: Weight Decay 추가 (GLM 4.5 paper 참고)
이유:
Overfitting 방지
GLM 4.5 paper에서 사용한 설정

**Comment by ryangkyung-thaki** [comment]
## 📊 실험 결과 업데이트 (2025-11-14)

### 실험 결과 요약

#### ✅ Test 1: Rank 16 + LR 1e-5
```yaml
lora_rank: 16
lora_alpha: 32
learning_rate: 1.0e-5
num_train_epochs: 1
data: 50k
```

**결과:**
| Checkpoint | KMMLU | Average Accuracy |
|-----------|-------|-----------------|
| Base Model | 0.200 | 0.373 |
| checkpoint-100 | 0.170 | 0.360 ⬇️ |
| checkpoint-200 | 0.170 | 0.353 ⬇️ |
| checkpoint-300 | 0.190 | 0.360 ⬇️ |
| checkpoint-391 | 0.190 | 0.357 ⬇️ |

❌ **문제점:** Base model보다 성능 하락, LR이 너무 높아서 불안정

---

#### ✅ Test 2: Rank 16 + LR 5e-6
```yaml
lora_rank: 16
lora_alpha: 32
learning_rate: 5.0e-6  # ← 감소
num_train_epochs: 1
data: 50k
```

**결과:**
| Checkpoint | KMMLU | Average Accuracy |
|-----------|-------|-----------------|
| Base Model | 0.200 | 0.373 |
| checkpoint-100 | 0.210 | 0.370 ✅ |
| checkpoint-200 | 0.180 | 0.363 ⬇️ |
| checkpoint-300 | 0.170 | 0.353 ⬇️ |
| checkpoint-391 | 0.190 | 0.357 ⬇️ |

💡 **개선:** 초반(100) 성능은 안정적이지만 여전히 전체적으로 하락

---

#### ✅ Test 3: Rank 32 + LR 5e-6 (Evaluation Results)
```yaml
lora_rank: 32
lora_alpha: 64
learning_rate: 5.0e-6
num_train_epochs: 1
data: 50k
```

**평가 결과:**
| Checkpoint | Average Accuracy |
|-----------|-----------------|
| Base Model | 0.373 |
| checkpoint-100 | 0.367 |
| checkpoint-200 | 0.357 |
| checkpoint-300 | 0.357 |
| checkpoint-391 | 0.357 |

❌ **결과:** Rank 16 → 32로 증가했으나 성능 개선 없음

---

### 🔍 핵심 발견: Overfitting 문제

**Training Loss 분석 (Rank 32, LR 5e-6):**
| Epoch | Eval Loss | 상태 |
|-------|-----------|------|
| 0.61 | 1.847 | ✅ **최저점** |
| 0.77 | 1.862 | ⬆️ 증가 시작 |
| 0.97 | 1.965 | ⬆️ 계속 증가 |
| 1.28 | 2.004 | ⬆️ Overfitting |
| 1.99 | 2.001 | ⬆️ Overfitting 지속 |

**Eval loss가 Epoch 0.61 이후 계속 증가** → **Overfitting 확인!**

---

## 🎯 문제 진단

### 1. **Learning Rate 문제 아님**
- 3.0e-4 → 1.0e-4 → 5.0e-5 → 1.0e-5 → 5.0e-6 모두 시도
- 5e-6에서 가장 안정적이지만 여전히 성능 하락

### 2. **Rank Capacity 문제 아님**
- Rank 8 → 16 → 32 시도
- Rank 증가해도 성능 개선 없음

### 3. **진짜 문제: Overfitting + 부족한 Regularization**
```
원인:
- Epoch 0.61 이후 eval loss 증가
- Training data만 외우고 일반화 실패
- Regularization 부족
```

---

## 🚀 새로운 실험 제안

### 실험 4: Anti-Overfitting 설정 (추천! ⭐⭐⭐)

```yaml
### LoRA 설정 (Rank 유지!)
lora_rank: 32  # 그대로 유지
lora_alpha: 64
lora_dropout: 0.1  # ← 추가!

### 학습 설정
per_device_train_batch_size: 32
gradient_accumulation_steps: 1
learning_rate: 3.0e-6  # 5e-6보다 더 낮게
num_train_epochs: 0.7  # ← 짧게! (최저점 근처에서 멈추기)
warmup_ratio: 0.1

### Regularization 강화
weight_decay: 0.2  # 0.1 → 0.2

### Early Stopping
eval_steps: 20
save_strategy: steps
save_steps: 20
load_best_model_at_end: true  # ← 추가!
metric_for_best_model: eval_loss
greater_is_better: false
```

**핵심 전략:**
1. ✅ **Early Stopping**: Epoch 0.7 근처에서 멈추기
2. ✅ **LoRA Dropout 추가**: Overfitting 방지
3. ✅ **Weight Decay 증가**: Regularization 강화
4. ✅ **Learning Rate 약간 감소**: 더 안정적인 학습
5. ❌ **Rank 증가 금지**: 이미 충분함, 더 올리면 overfitting 악화

---

## 📌 결론

1. **Learning Rate를 낮추는 것만으로는 부족** (이미 5e-6까지 시도)
2. **Rank를 올리는 것도 효과 없음** (32까지 시도)
3. **진짜 문제는 Overfitting** (Eval loss가 epoch 0.61 이후 증가)
4. **해결책: Regularization 강화 + Early Stopping**

**다음 실험: Rank 32 유지 + Dropout 추가 + Weight Decay 증가 + Epoch 0.7로 제한**
